pkg/broker/user_info.go

```diff
@@ -0,0 +1,62 @@
+package broker
+
+import (
+	"encoding/json"
+	"fmt"
+
+	osb "github.com/pmorie/go-open-service-broker-client/v2"
+)
+
+// ParseKubernetesIdentity - creates a kubernetes identity from the
+// orginating identity
+func ParseKubernetesIdentity(o osb.OriginatingIdentity) (*K8SIdentity, error) {
```

Comment:
I would prefer to have this as a generic `ParseIdentity` function instead, so as a user I don't have to match platforms but that the lib resolves that for me and returns the correct Identity.

But can't currently come up with any sane suggestion, any ideas?

Reply:
I thought of that as well, the issue that I keep coming up with is that you would need an interface to encapsulate the behavior between the two Identities. And they are different as when I get the k8s identity I want the UserInfo structure, and when I get cloud foundry I expect a map[string]interface{}. 

We could eventually do something like `GetValue(valueName string) interface{}` or `GetValueString(valueName) (string, error)` but I think that will be very complex and maybe less usable then dealing with calling different methods? 

@pmorie I think the above point is what you thought as well when we discussed my first attempt at this PR.

Reply:
We could have a union type `Identity` with fields `Kubernetes` and `CloudFoundry` and then an `Unknown` bucket for everything else.

Reply:
So `Kubernetes` = `UserInfo`, `CloudFoundry` = `map[string]interface{}`? So is unkown for a platform that we do not understand?

Reply:
Perhaps you could parse the identity and return a wrapped result that can return the type, and has known named methods to return typed k8s and CF identity objects, and maybe a method to also return the data back as a map? So the caller has to do some work to figure out which representation to ask for but it will get a typed object back without the weird union of the three things. 

so maybe like:

```
func ParseIdentity(o osb.OriginatingIdentity) (*Identity, error)

identity, err := ParseIdentity(osbIdent)

switch identity.Type {
case K8s: 
    k8sIdentity := identity.AsK8s()
case CF:
    cfIdentity := identity.AsCF()  
case Unknown:
    mapIdentity := identity.AsMap()
}

...
```

Reply:
I am fine with this approach if it sounds better? @LiliC @pmorie

Reply:
The approach @n3wscott suggested sgtm 👍

Reply:
> a quick question about that implementation what should the AsXX function do if it is called but the platform type does not match? should it return the zero value s for the typed struct? (or map)?

Can you rephrase that, not sure I understood? You mean if it's neither k8s nor CF?

> Should the validation of the type (CF check for key "user_id") be in the AsCF function or do you think that the ParseIdentity does that? Sorry just want to make sure everyone is on the same page for the next implementation of this one.

IMO I would return early so in the `ParseIdentity()` function already we should verify each platform before returning it, so in that case it would just return the error.

Reply:
> IMO I would return early so in the ParseIdentity() function already we should verify each platform before returning it, so in that case it would just return the error.

I agree with this. Internally it knows what the identity type is (because it also supports .Type, so it can throw an error if Type is k8s but you are asking for CF.)

Reply:
Btw, I am short handing the named for the As**, feel free to use AsKubernetes() etc.

Reply:
I will implement this way if you want, just know the way it is now is a simpler way for a caller to interact IMO then what we are proposing EX:

```golang
switch originatingIdentity.Platform {
case osb.KubernetesPlatform:
     k8sIdentity := ParseKubernetesIdentity(originatingIdentity)
     ....
case osb.CloudFoundryPlatform:
    cfIdentity := ParseCloudFoundryIdentity(originatingIdentity)
default:
    //decode to map[string]interface{} if you want.
     .....
}
```

VS
```
func ParseIdentity(o osb.OriginatingIdentity) (*Identity, error) {...}

identity, err := ParseIdentity(osbIdent)

switch identity.Type {
case K8s: 
    k8sIdentity := identity.AsK8s()
case CF:
    cfIdentity := identity.AsCF()  
case Unknown:
    mapIdentity := identity.AsMap()
}
```

I personally don't see the value in adding `identity.Type` and its constants. Adding an Identity struct attaching three new methods and having a client still have to do a switch statement just after making a call when the above would work. Just my $0.02

Reply:
If there were no switch in the second case I could see the value in it. But with both needing a case statement, I'm not sure I see why one is easier than the other. You still need to know am I running in k8s or cf.

Reply:
I think the former is cleaner, but only because the assumption here is that the processing was actually already done to get us the `osb.OriginatingIdentity` resource. If you're able to switch off `originatingIdentity.Platform`, there's no need to further process it into a generic `Identity` struct to then switch off of.

Reply:
Use a union type; a struct that has pointers to a k8s identity, cf identity, and other.